In [18]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt


In [4]:
df=pd.read_csv("C:/Users/Samarth Thakur/Downloads/01Exercise1.csv")
df

gender married   ch  income  loanamt status
0      Male      No  1.0    5849      NaN      Y
1      Male     Yes  1.0    4583    128.0      N
2      Male     Yes  1.0    3000     66.0      Y
3      Male     Yes  1.0    2583    120.0      Y
4      Male      No  1.0    6000    141.0      Y
..      ...     ...  ...     ...      ...    ...
609  Female      No  1.0    2900     71.0      Y
610    Male     Yes  1.0    4106     40.0      Y
611    Male     Yes  1.0    8072    253.0      Y
612    Male     Yes  1.0    7583    187.0      Y
613  Female      No  0.0    4583    133.0      N

[614 rows x 6 columns]

In [5]:
df.dtypes

gender      object
married     object
ch         float64
income       int64
loanamt    float64
status      object
dtype: object

In [6]:
loandf=df.copy()
loandf

gender married   ch  income  loanamt status
0      Male      No  1.0    5849      NaN      Y
1      Male     Yes  1.0    4583    128.0      N
2      Male     Yes  1.0    3000     66.0      Y
3      Male     Yes  1.0    2583    120.0      Y
4      Male      No  1.0    6000    141.0      Y
..      ...     ...  ...     ...      ...    ...
609  Female      No  1.0    2900     71.0      Y
610    Male     Yes  1.0    4106     40.0      Y
611    Male     Yes  1.0    8072    253.0      Y
612    Male     Yes  1.0    7583    187.0      Y
613  Female      No  0.0    4583    133.0      N

[614 rows x 6 columns]

In [7]:
df.shape

(614, 6)

In [8]:
df.describe()

ch        income     loanamt
count  564.000000    614.000000  592.000000
mean     0.842199   5403.459283  146.412162
std      0.364878   6109.041673   85.587325
min      0.000000    150.000000    9.000000
25%      1.000000   2877.500000  100.000000
50%      1.000000   3812.500000  128.000000
75%      1.000000   5795.000000  168.000000
max      1.000000  81000.000000  700.000000

In [14]:
loandf.corr()

ch    income   loanamt  gender_Male  married_Yes  status_Y
ch           1.000000 -0.035671 -0.021177     0.010878     0.016856  0.559912
income      -0.035671  1.000000  0.524141     0.046995     0.045983 -0.027201
loanamt     -0.021177  0.524141  1.000000     0.103965     0.171934 -0.052089
gender_Male  0.010878  0.046995  0.103965     1.000000     0.363102  0.048585
married_Yes  0.016856  0.045983  0.171934     0.363102     1.000000  0.094535
status_Y     0.559912 -0.027201 -0.052089     0.048585     0.094535  1.000000

In [10]:
loandf.isnull().sum()

gender     13
married     3
ch         50
income      0
loanamt    22
status      0
dtype: int64

In [11]:
loandf=loandf.dropna()
loandf.shape

(529, 6)

In [12]:
loandf.isnull().sum()

gender     0
married    0
ch         0
income     0
loanamt    0
status     0
dtype: int64

In [17]:
df.corr()

ch    income   loanamt
ch       1.000000 -0.014715 -0.008433
income  -0.014715  1.000000  0.570909
loanamt -0.008433  0.570909  1.000000

In [16]:
loandf.corr()

ch    income   loanamt  gender_Male  married_Yes  status_Y
ch           1.000000 -0.035671 -0.021177     0.010878     0.016856  0.559912
income      -0.035671  1.000000  0.524141     0.046995     0.045983 -0.027201
loanamt     -0.021177  0.524141  1.000000     0.103965     0.171934 -0.052089
gender_Male  0.010878  0.046995  0.103965     1.000000     0.363102  0.048585
married_Yes  0.016856  0.045983  0.171934     0.363102     1.000000  0.094535
status_Y     0.559912 -0.027201 -0.052089     0.048585     0.094535  1.000000

In [15]:
loandf=pd.get_dummies(loandf,drop_first=True)
loandf.head()

ch  income  loanamt  gender_Male  married_Yes  status_Y
1  1.0    4583    128.0            1            1         0
2  1.0    3000     66.0            1            1         1
3  1.0    2583    120.0            1            1         1
4  1.0    6000    141.0            1            0         1
5  1.0    5417    267.0            1            1         1

In [10]:
#normalize the data for loanamt and income using standardscaler
from sklearn.preprocessing import StandardScaler
scalar=StandardScaler()

loandf['income']=scalar.fit_transform(loandf[['income']])
loandf['loanamt']=scalar.fit_transform(loandf[['loanamt']])

In [11]:
loandf.head()

ch    income   loanamt  gender_Male  married_Yes  status_Y
1  1.0 -0.128073 -0.194250            1            1         0
2  1.0 -0.392077 -0.971015            1            1         1
3  1.0 -0.461621 -0.294478            1            1         1
4  1.0  0.108246 -0.031380            1            0         1
5  1.0  0.011017  1.547205            1            1         1

# Building the model 

In [12]:
Y=loandf[['status_Y']]
X=loandf.drop(['status_Y'],axis=1)

In [13]:
from sklearn.model_selection import train_test_split

X_train,X_test,Y_train,Y_test=\
    train_test_split(X,Y,test_size=0.3,random_state=1234,stratify=Y)

In [14]:
from sklearn.linear_model import LogisticRegression
lr=LogisticRegression()
lr.fit(X_train,Y_train)

E:\python\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression()

In [15]:
Y_predict=lr.predict(X_test)
print(Y_predict)

[0 0 1 1 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 0 1 1 1 0 1
 1 0 1 1 1 1 1 0 0 1 1 1 1 0 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 0 1 1 1 1 1 1 0 1 1 1 0 0 1 1 1 1 1 1 1 0 1 0 1 1 1 1 1 1 1 1
 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 0 1 1 0 1 1 1 0 0 1 1 1 1 1 1 1 1 1 1 1 0 1
 1 1 1 0 1 1 1 1 1 1 1]


In [16]:
#building confusion matrix
from sklearn.metrics import confusion_matrix

cm=confusion_matrix(Y_test,Y_predict)
cm

array([[ 29,  20],
       [  2, 108]], dtype=int64)

we find that we have correctly predicted that 29 loans are rejected and 108 are approved also 20 loans are actually rejejcted but we approved them and for 2  that needs to approved we rejected overall it is not bad model because out of 159 we correctly predited of 137(108+29)

In [17]:
score=lr.score(X_test,Y_test)
score
#aroun 86% accuracy

0.8616352201257862